File: packages/typescript/jest-environment/src/environment/dogu-config.ts
import fs from 'fs';
import JSON5 from 'json5';
import { validate } from 'jsonschema';
import _ from 'lodash';
import path from 'path';
import url from 'url';
import { createLogger } from './common-utils.js';

/**
 * @note "tsc" does not support import.meta. The transfile is in babel and the tsc generates only the type.
 */
/* @ts-ignore */
const ConfigDirPath = path.resolve(path.dirname(url.fileURLToPath(import.meta.url)));
const SchemaFilePath = path.resolve(ConfigDirPath, 'dogu.config.schema.json');
const ConfigFilePath = path.resolve(process.cwd(), 'dogu.config.json');
const ApiBaseUrlPattern = /^(https?):\/\/([^:\/\s]+)(:([0-9]+))?\/?/i;

async function loadConfigFileSchema(): Promise<object> {
  const content = await fs.promises.readFile(SchemaFilePath, 'utf8');
  const schema = JSON5.parse(content) as object;
  return schema;
}

async function loadConfigFile(): Promise<object> {
  const content = await fs.promises.readFile(ConfigFilePath, 'utf8');
  const config = JSON5.parse(content) as object;
  return config;
}

async function validateConfigFile(config: object, schema: object): Promise<void> {
  const validationResult = validate(config, schema);
  if (!validationResult.valid) {
    throw new Error(`Invalid config: ${JSON.stringify(validationResult.errors)}`);
  }
}

async function loadAndValidateConfigFile(): Promise<object> {
  const schema = await loadConfigFileSchema();
  const config = await loadConfigFile();
  await validateConfigFile(config, schema);
  return config;
}

async function isConfigFileExist(): Promise<boolean> {
  const isConfigFileExist = await fs.promises.stat(ConfigFilePath).catch(() => null);
  return !!isConfigFileExist;
}

export class DoguConfig {
  private readonly logger = createLogger('DoguConfig');

  constructor(private readonly config: object) {}

  get version(): number {
    return _.get(this.config, 'version') as unknown as number;
  }

  get apiBaseUrl(): string {
    return process.env.DOGU_API_BASE_URL || (_.get(this.config, 'apiBaseUrl') as unknown as string);
  }

  get organizationId(): string {
    return process.env.DOGU_ORGANIZATION_ID || (_.get(this.config, 'organizationId') as unknown as string);
  }

  get projectId(): string {
    return process.env.DOGU_PROJECT_ID || (_.get(this.config, 'projectId') as unknown as string);
  }

  get token(): string {
    return process.env.DOGU_TOKEN || (_.get(this.config, 'token') as unknown as string);
  }

  get runsOn(): string | string[] {
    return process.env.DOGU_RUNS_ON || (_.get(this.config, 'runsOn') as unknown as string | string[]);
  }

  get browserName(): string | undefined {
    return process.env.DOGU_BROWSER_NAME || _.get(this.config, 'browserName');
  }

  get browserVersion(): string | undefined {
    return process.env.DOGU_BROWSER_VERSION || _.get(this.config, 'browserVersion');
  }

  get appVersion(): string | undefined {
    return process.env.DOGU_APP_VERSION || _.get(this.config, 'appVersion');
  }

  get deviceId(): string | undefined {
    return process.env.DOGU_DEVICE_ID;
  }

  get stepId(): string | undefined {
    return process.env.DOGU_STEP_ID;
  }

  get hostToken(): string | undefined {
    return process.env.DOGU_HOST_TOKEN;
  }

  get failFast(): boolean {
    return _.get(this.config, 'failFast', false);
  }

  parseApiBaseUrl(): { protocol: string; hostname: string; port: number } {
    const { apiBaseUrl } = this;
    const matches = apiBaseUrl.match(ApiBaseUrlPattern);
    if (!matches) {
      throw new Error(`Invalid apiBaseUrl: ${apiBaseUrl}`);
    }
    const [, protocol, hostname, , port] = matches;
    if (typeof protocol !== 'string' || typeof hostname !== 'string' || (port && typeof port !== 'string')) {
      throw new Error(`Internal error: ${apiBaseUrl} is not matched`);
    }
    if (port) {
      const portNumber = parseInt(port, 10);
      if (Number.isNaN(portNumber)) {
        throw new Error(`Invalid port: ${port}`);
      }
      return { protocol, hostname, port: portNumber };
    }
    return { protocol, hostname, port: protocol === 'https' ? 443 : 80 };
  }
}

export class DoguConfigFactory {
  private readonly logger = createLogger('DoguConfigFactory');

  async create(): Promise<DoguConfig> {
    if (await isConfigFileExist()) {
      const config = await loadAndValidateConfigFile();
      this.logger.info('dogu.config.json is loaded');
      return new DoguConfig(config);
    }
    this.logger.info('dogu.config.json is not found. use default config');
    return new DoguConfig({});
  }
}

File: packages/python/pytest-dogu-sdk/src/pytest_dogu_sdk/dogu_config.py
from typing import List, Optional, Union
import os

from pyjson5 import decode_io  # pylint: disable=no-name-in-module
from jsonschema import validate


_config_dir_path = os.path.dirname(os.path.realpath(__file__))
_schema_file_path = os.path.join(_config_dir_path, "dogu.config.schema.json")
_config_file_path = os.path.join(os.getcwd(), "dogu.config.json")


def _load_config_file_schema() -> dict:
    with open(_schema_file_path, "r", encoding="utf8") as file:
        return decode_io(file)


def _load_config_file() -> dict:
    with open(_config_file_path, "r", encoding="utf8") as file:
        return decode_io(file)


def _validate_config_file(config: dict, schema: dict) -> None:
    validate(instance=config, schema=schema)


def _load_and_validate_config_file() -> dict:
    schema = _load_config_file_schema()
    config = _load_config_file()
    _validate_config_file(config, schema)
    return config


def _is_config_file_exist() -> bool:
    return os.path.exists(_config_file_path)


class DoguConfig:
    def __init__(self, config: dict):
        self._config = config

    @property
    def version(self) -> int:
        return self._config["version"]

    @property
    def api_base_url(self) -> str:
        api_base_url = os.environ.get("DOGU_API_BASE_URL", self._config.get("apiBaseUrl", ""))
        if api_base_url.endswith("/"):
            return api_base_url[:-1]
        return api_base_url

    @property
    def organization_id(self) -> str:
        return os.environ.get("DOGU_ORGANIZATION_ID", self._config.get("organizationId", ""))

    @property
    def project_id(self) -> str:
        return os.environ.get("DOGU_PROJECT_ID", self._config.get("projectId", ""))

    @property
    def token(self) -> str:
        return os.environ.get("DOGU_TOKEN", self._config["token"])

    @property
    def runs_on(self) -> Union[str, List[str]]:
        return os.environ.get("DOGU_RUNS_ON", self._config["runsOn"])

    @property
    def browser_name(self) -> Optional[str]:
        return os.environ.get("DOGU_BROWSER_NAME", self._config.get("browserName"))

    @property
    def browser_version(self) -> Optional[str]:
        return os.environ.get(
            "DOGU_BROWSER_VERSION", self._config.get("browserVersion")
        )

    @property
    def app_version(self) -> Optional[str]:
        return os.environ.get("DOGU_APP_VERSION", self._config.get("appVersion"))

    @property
    def device_id(self) -> Optional[str]:
        return os.environ.get("DOGU_DEVICE_ID")

    @property
    def step_id(self) -> Optional[str]:
        return os.environ.get("DOGU_STEP_ID")

    @property
    def host_token(self) -> Optional[str]:
        return os.environ.get("DOGU_HOST_TOKEN")

    @property
    def fail_fast(self) -> bool:
        return self._config.get("failFast", False)


class DoguConfigFactory:
    def create(self) -> DoguConfig:
        if _is_config_file_exist():
            config_file_dict = _load_and_validate_config_file()
            print("[dogu] dogu.config.json is loaded")
            return DoguConfig(config_file_dict)
        print("[dogu] dogu.config.json is not found. use default config")
        return DoguConfig({})
